Removes the "Index contains time-based events" checkbox

ycombinator opened the pull request:
This PR addresses the first part of #10444.

Currently we show a checkbox labeled "Index contains time-based events". The user has to check this box, and is then shown a select list with time fields.

This PR simplifies this UX by:
* Taking away the checkbox and always showing the select list
* For valid index patterns, showing the first item in the select list as "No Time Filter field". For empty or non-existing index patterns, the select list is empty
* For index patterns with no time fields, disabling the select list
* For index patterns with time fields, enabling the select list
* Renaming the label of the select list to "Time filter field name" to better reflect where the selected time field is used
* [Cleanup in related code] Renaming the label for the time field on the Edit Index Pattern page to "Time filter field name" for consistency

## When no index pattern is specified

<img width="1082" alt="screen shot 2017-04-25 at 6 52 38 am" src="https://cloud.githubusercontent.com/assets/51061/25392461/89906112-298d-11e7-9975-9ee8635f714b.png">

## When a non-existing index name/pattern is specified

<img width="1085" alt="screen shot 2017-04-25 at 6 52 57 am" src="https://cloud.githubusercontent.com/assets/51061/25392453/86981f7c-298d-11e7-9531-3665ebc642b6.png">

## When an index containing no time fields is specified

<img width="1084" alt="screen shot 2017-04-25 at 6 53 07 am" src="https://cloud.githubusercontent.com/assets/51061/25392429/79fc2236-298d-11e7-8299-f138fb511710.png">

## When an index containing time fields is specified

<img width="1085" alt="screen shot 2017-04-25 at 6 53 40 am" src="https://cloud.githubusercontent.com/assets/51061/25392418/73192da6-298d-11e7-8ffb-f7f599e44cf1.png">

<img width="1084" alt="screen shot 2017-04-25 at 6 53 49 am" src="https://cloud.githubusercontent.com/assets/51061/25392412/6f43123c-298d-11e7-970a-eea86f78f8cd.png">

ycombinator commented:
/cc @skearns64

epixa commented:
I assume you wanted the v6.0.0 label here :)

epixa commented:
Two suggestions:

Can you move the "Unable to fetch mappings" error message to immediately below the index pattern field?  This way the error appears at the place that caused it/where it can be fixed.  You may need to play around with a disabled submit button instead of hiding the submit button entirely, but honestly this is really how the form should work anyway.  Not showing a submit button until everything is kosher is kind of awkward.  I don't want to churn too much on this, so if this seems to blow open the issue, we can address it in a follow up.

Rather than "No Time Filter Field", which doesn't do a great job describing why that may be the case or what the effect may be, how about two different possibilities:

For indices without time fields at all: `Not time-based: Matching indices do not have date fields`

For indices with time fields but without one selected: `Not time-based: No date field selected as the primary time field`

Thoughts?

ycombinator commented:
> Can you move the "Unable to fetch mappings" error message to immediately below the index pattern field?  This way the error appears at the place that caused it/where it can be fixed.  You may need to play around with a disabled submit button instead of hiding the submit button entirely, but honestly this is really how the form should work anyway.  Not showing a submit button until everything is kosher is kind of awkward.  I don't want to churn too much on this, so if this seems to blow open the issue, we can address it in a follow up.

Yeah, I agree with this change but I'd like to address it in a follow up PR. I've created an issue so we don't forget about it: https://github.com/elastic/kibana/issues/11410.

> Rather than "No Time Filter Field", which doesn't do a great job describing why that may be the case or what the effect may be, how about two different possibilities:
>
> For indices without time fields at all: `Not time-based: Matching indices do not have date fields`
>
> For indices with time fields but without one selected: `Not time-based: No date field selected as the primary time field`

Yeah, ++ to making the messages more helpful to the user. Changing.

cjcenizal commented:
## When an index containing no time fields is specified

<img width="1084" alt="screen shot 2017-04-25 at 6 53 07 am" src="https://cloud.githubusercontent.com/assets/51061/25392429/79fc2236-298d-11e7-8299-f138fb511710.png">

In this situation, can we replace the select with some text stating, "The indices which match this index pattern don't contain any time fields."

## When an index containing time fields is specified

<img width="1085" alt="screen shot 2017-04-25 at 6 53 40 am" src="https://cloud.githubusercontent.com/assets/51061/25392418/73192da6-298d-11e7-8ffb-f7f599e44cf1.png">

Can we change "No time filter field" to "None selected"? The former sounds like there simply are no time filter fields available, and I think the latter clarifies that the user hasn't picked one yet.

ycombinator commented:
@cjcenizal Did you see Court's comment earlier? It addresses the same issues you brought up. I'm putting up a new commit with some better verbiage shortly.

cjcenizal commented:
@ycombinator Yeah I saw his comment afterwards and edited my comment. We have the same things in mind, but I think the language I suggested is a little more human-readable.

ycombinator commented:
Okay, changing to the language suggested by CJ.

cjcenizal commented:
Winning!

ycombinator commented:
I updated the verbiage per @cjcenizal's suggestions but I kept the select lists always in place per the spirit of the conversation here (or my understanding of it anyway): https://github.com/elastic/kibana/issues/10444#issuecomment-280788995.

Updated screenshots for all the cases below.

## When no index pattern is specified

<img width="1085" alt="screen shot 2017-04-25 at 9 38 17 am" src="https://cloud.githubusercontent.com/assets/51061/25397419/176f38dc-299d-11e7-882c-9dc318fbba03.png">

## When a non-existing index name/pattern is specified


<img width="1084" alt="screen shot 2017-04-25 at 9 38 31 am" src="https://cloud.githubusercontent.com/assets/51061/25397415/146ca8ea-299d-11e7-9633-14096bec3815.png">

## When an index containing no time fields is specified

<img width="1085" alt="screen shot 2017-04-25 at 9 38 45 am" src="https://cloud.githubusercontent.com/assets/51061/25397411/11aa6db8-299d-11e7-9590-d95579fc43c7.png">

## When an index containing time fields is specified

<img width="1083" alt="screen shot 2017-04-25 at 9 37 49 am" src="https://cloud.githubusercontent.com/assets/51061/25397417/17378950-299d-11e7-9e09-90471478107b.png">

<img width="1083" alt="screen shot 2017-04-25 at 9 38 03 am" src="https://cloud.githubusercontent.com/assets/51061/25397418/174c7b30-299d-11e7-9eac-5dcd4a547798.png">

cjcenizal commented:
@ycombinator What do you think of this?

![index_pattern_creation_time_fields](https://cloud.githubusercontent.com/assets/1238659/25399283/cc3546d4-29a3-11e7-85fc-37ee6fc2b97b.gif)

If you think this works, we can implement it by adding a paragraph below the select:

```html
<p class="help-block" ng-if="!controller.doesIndexHaveDateFields">
  The indices which match this index pattern don&rsquo;t contain any time fields.
</p>
```

And updating the controller like this:

```javascript
if (this.dateFields.length === 0) {
  this.doesIndexHaveDateFields = false;
  this.dateFields.unshift(TIME_FILTER_FIELD_OPTIONS.NO_DATE_FIELDS_IN_INDEX);
} else {
  this.doesIndexHaveDateFields = true;
  this.dateFields.unshift(TIME_FILTER_FIELD_OPTIONS.NO_DATE_FIELD_SELECTED);
}
```

And making this change in the translation file:

```json
  "KIBANA-NO_DATE_FIELD_SELECTED": "None selected",
  "KIBANA-NO_DATE_FIELDS_IN_INDEX": "None available",
```

cjcenizal commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern/create_index_pattern.html:
```diff
@@ -41,19 +41,9 @@
+          <div class="form-group">
             <label>
-              <span translate="KIBANA-TIME_FIELD_NAME"></span>
+              <span translate="KIBANA-TIME_FILTER_FIELD_NAME"></span>
```
Just curious: what does time filter mean in this context?

ycombinator replied:
I'm using that per this suggestion: https://github.com/elastic/kibana/issues/10444#issuecomment-280773735.

I think "time picker" might be more appropriate but I'm not sure if users know what that is.

cjcenizal replied:
Hmmm... I guess "time filter" works. I like "time picker" because I know what that is, but I'm afraid that might couple these field too closely to a particular UI component.

cjcenizal commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern/create_index_pattern.js:
```diff
@@ -40,9 +39,18 @@ uiModules.get('apps/management')
+      name: $translate.instant('KIBANA-NO_DATE_FIELD_SELECTED')
+    },
+    NO_DATE_FIELDS_IN_INDEX: {
+      name: $translate.instant('KIBANA-NO_DATE_FIELDS_IN_INDEX')
```
Really minor suggestion: technically, we're referring to "indices", and not a single index, right? It all depends on how many indices match the index pattern. Maybe it's just me, but when I first started using Kibana, it took awhile for me to understand the difference between index pattern and index. I think renaming these to be plural will help reinforce the relationship, e.g. `NO_DATE_FIELDS_IN_INDICES` and `KIBANA-NO_DATE_FIELDS_IN_INDICES`.

cjcenizal commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern/create_index_pattern.html:
```diff
@@ -41,19 +41,9 @@
-          <div class="form-group time-and-pattern">
-            <label>
-              <input
-                data-test-subj="createIndexPatternIsTimeBasedCheckBox"
```
We'll need to update the functional tests to remove references to this element.

cjcenizal commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern/create_index_pattern.js:
```diff
@@ -184,14 +203,15 @@ uiModules.get('apps/management')
-      : undefined;
+    let timeFieldName;
+    if ((this.newIndexPattern.timeField !== TIME_FILTER_FIELD_OPTIONS.NO_DATE_FIELD_SELECTED)
+      && (this.newIndexPattern.timeField !== TIME_FILTER_FIELD_OPTIONS.NO_DATE_FIELDS_IN_INDEX)) {
```
I think this could be reduced to:

```javascript
![
  TIME_FILTER_FIELD_OPTIONS.NO_DATE_FIELD_SELECTED,
  TIME_FILTER_FIELD_OPTIONS.NO_DATE_FIELDS_IN_INDEX
].includes(this.newIndexPattern.timeField)
```

ycombinator replied:
Yeah, I considered something like that but I find it less readable. In particular, I found the `.includes` relationship to be inverted (in terms of readability).

I'd be fine with shortening this if we could express this in reverse, so something like (pseudocode):

`!this.newIndexPattern.timeField.isOneOf([ ... ])`

But I don't think there's any way to do that, either in JS itself or using a helper library.

So I'd prefer to leave it the way it-is for added readability.

If you strongly disagree, we can get a third person to break the tie.

cjcenizal replied:
It's all good, I don't feel very strongly about this.

ycombinator commented:
@cjcenizal This is ready for review again, based on your feedback. I still need to implement the spinner though, so if you just want to wait till that's done, that's cool too.

ycombinator commented:
@cjcenizal On second thought, I'm going to tackle the spinner as a separate PR, since there's a neat little separate issue for it that specs it out well already: https://github.com/elastic/kibana/issues/10439.

archanid commented:
@cjcenizal @ycombinator Just being a ceiling cat here, but in the gif that CJ posted, I couldn't tell that the text `The indices which match this index pattern don't contain any time fields.` appeared below as a message corresponding to the change in the input box. Maybe flashing a background color highlight that fades after a second would help draw attention?

w33ble commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern/create_index_pattern.html:
```diff
@@ -63,13 +53,16 @@
             <select
               data-test-subj="createIndexPatternTimeFieldSelect"
-              ng-disabled="controller.fetchFieldsError"
+              ng-disabled="controller.fetchFieldsError || (controller.dateFields.length === 1)"
```
Might be clearer to use `!controller.doesIndexHaveDateFields` here.

ycombinator replied:
Ah yes, that was introduced later and should definitely be used here. Thanks!

w33ble commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern/create_index_pattern.js:
```diff
@@ -89,7 +97,18 @@ uiModules.get('apps/management')
+    }
+
+    this.dateFields = results.dateFields || [];
+    this.doesIndexHaveDateFields = this.dateFields.length > 0;
```
`this.indexHasDateFields` fits the pattern better, and reads better when used for conditionals.

cjcenizal commented:
@archanid Good point! I think it's better to address that as part of our new notifications designs. Would you mind chiming in with your thoughts and cross-linking on https://github.com/elastic/kibana/issues/8194 ?

ycombinator commented:
Tests are failing but they are unrelated to the changes in this PR. Merging...

ycombinator commented:
Backported to:

* `5.x` / `5.5`: 7aae24b
